libraries/AP_RCTelemetry/AP_Spektrum_Telem.cpp

```diff
@@ -0,0 +1,618 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+/*
+   Spektrum Telemetry library, based on AP_Frsky_Telem.cpp
+   See https://www.spektrumrc.com/ProdInfo/Files/SPM_Telemetry_Developers_Specs.pdf
+*/
+
+#include "AP_Spektrum_Telem.h"
+#include <AP_HAL/utility/sparse-endian.h>
+#include <AP_AHRS/AP_AHRS.h>
+#include <AP_RPM/AP_RPM.h>
+#include <AP_Airspeed/AP_Airspeed.h>
+#include <AP_Compass/AP_Compass.h>
+#include <AP_BattMonitor/AP_BattMonitor.h>
+#include <AP_RangeFinder/AP_RangeFinder.h>
+#include <AP_Common/AP_FWVersion.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Common/Location.h>
+#include <AP_GPS/AP_GPS.h>
+#include <AP_Baro/AP_Baro.h>
+#include <AP_GPS/AP_GPS.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_RTC/AP_RTC.h>
+#ifdef HAVE_AP_BLHELI_SUPPORT
+#include <AP_BLheli/AP_BLHeli.h>
+#endif
+#include <math.h>
+
+#if HAL_SPEKTRUM_TELEM_ENABLED
+
+#define MICROSEC_PER_MINUTE 60000000
+#define MAX_TEXTGEN_LEN     13
+
+//#define SPKT_DEBUG
+#ifdef SPKT_DEBUG
+# define debug(fmt, args...)	hal.console->printf("SPKT:" fmt "\n", ##args)
+#else
+# define debug(fmt, args...)	do {} while(0)
+#endif
+
+extern const AP_HAL::HAL& hal;
+
+AP_Spektrum_Telem *AP_Spektrum_Telem::singleton;
+
+AP_Spektrum_Telem::AP_Spektrum_Telem() : AP_RCTelemetry(0)
+{
+    singleton = this;
+}
+
+AP_Spektrum_Telem::~AP_Spektrum_Telem(void)
+{
+    singleton = nullptr;
+}
+
+bool AP_Spektrum_Telem::init(void)
+{
+    return AP_RCTelemetry::init();
```

Comment:
this init function needs to fail if we don't have a serial port setup for RC telemetry. At the moment it returns true unconditionally. That means we're paying for the AP_Spektrum_Telem object on all boards all the time, even if there is no uart to associate with it.
See how it is done for the frsky telem instance, which checks for uarts

Reply:
It's not quite as easy as that - we only require a UART to support RCIN in this instance, so that is all we can check for. It's not like FrSky where you separately need to specify the telemetry protocol. I have added the check is that enough?

Reply:
Still not quite right, sorry. Spektrum telem is in the same situation as FPort, where there is no parameter to indicate 100% that it should be created. You have the code in place in AP_Spektrum_Telem::get_telem_data() to auto-create on demand, so I think we can just remove the code in GCS_MAVLink to create on startup, then it will get created the first time we ask for telem data, which means we only create it when we know we really need it

Reply:
We still need to write the messages in GCS_Common.cpp so I have made this only happen if Spektrum telemetry has been created. This works well (removing all the mavlink code does not).